### PR TITLE
SALTO-5124: Deploy Jira NotificationScheme type

### PR DESCRIPTION
**Pull request by shir-reifenberg**
Our current deployment of NotificationScheme no longer works due to changes in the private API

---

This PR replaces the API used for deployment of `NotificationScheme` from private API deployment (using JSP endpoint) to official `NotificationScheme` API.

unit test are still missing, will add soon

---
_Release Notes_: 

_Jira_adapter_:
- Fix bug in deployment of NotificationScheme type

---
_User Notifications_: 
None

### CODE REVIEW

File: packages/jira-adapter/src/config/api_config.ts [hunk 1]
```diff
@@ -2410,7 +2445,8 @@ const SUPPORTED_TYPES = {
 
 export const DEFAULT_API_DEFINITIONS: JiraApiConfig = {
   platformSwagger: {
-    url: 'https://raw.githubusercontent.com/salto-io/adapter-swaggers/main/jira/platform-swagger.v3.json',
+    url: 'https://raw.githubusercontent.com/salto-io/adapter-swaggers/8f9045fb5bbc528cead1741820c59ae9d60720d1/jira/platform-swagger.v3.json',
+    // url: 'https://raw.githubusercontent.com/salto-io/adapter-swaggers/main/jira/platform-swagger.v3.json',
```
Comment (shir-reifenberg) [inline packages/jira-adapter/src/config/api_config.ts:2449]: shouldn't be merged (same for line 2594)

File: packages/jira-adapter/src/filters/notification_scheme/notification_events.ts [hunk 2]
```diff
@@ -109,47 +82,42 @@ export const transformAllNotificationEvents = (notificationSchemeValues: Values)
...
       eventType,
-      convertValuesToJSPBody({
-        ...event,
-        name: getEventKey(event),
-        id: instance.value.notificationIds?.[getEventKey(event)],
-      }, instance),
-    ))
+      {
+        notificationSchemeEvents: [{ event: eventEntry.event, notifications: [eventEntry.notification] }],
+        // notificaitonIds might not exist
```
Comment (omrilit) [inline packages/jira-adapter/src/filters/notification_scheme/notification_events.ts:116]: should we log when we encounter such?

File: packages/jira-adapter/src/filters/notification_scheme/notification_events.ts [hunk 3]
```diff
@@ -164,19 +132,25 @@ const getEventType = async (change: Change<InstanceElement>): Promise<ObjectType
   return eventType
 }
 
-const getEventChanges = async (
-  change: AdditionChange<InstanceElement> | ModificationChange<InstanceElement>
+export const getEventChanges = async (
```
Comment (omrilit) [inline packages/jira-adapter/src/filters/notification_scheme/notification_events.ts:135]: when this function is exported I think that it should have a more self-explanatory name, WDYT?

File: packages/jira-adapter/src/filters/notification_scheme/notification_scheme_deployment.ts [hunk 4]
```diff
@@ -13,144 +13,210 @@
...
+
+const getDeployableNotificationEvent = (eventEntries: NotificationEventValue): NotificationEvent[] =>
+  eventEntries.map(eventEntry =>
+    ({
+      event: {
+        id: eventEntry.eventType,
+      },
+      notifications: eventEntry.notifications?.map(notification => ({
+        notificationType: notification.type,
+        ...notification,
```
Comment (omrilit) [inline packages/jira-adapter/src/filters/notification_scheme/notification_scheme_deployment.ts:62]: Is it fine to have here both `notificationType` & `type`? or should we be more explicit with the transformation here?

File: packages/jira-adapter/src/filters/notification_scheme/notification_scheme_deployment.ts [hunk 5]
```diff
@@ -13,144 +13,210 @@
...
+      Joi.object({
+        type: Joi.string().required(),
+      }).unknown(true)
+    ).optional(),
+  }).unknown(true)
+)
+
+const isNotificationEventValue = createSchemeGuard<NotificationEventValue>(NOTIFICATION_EVENT_VALUE_SCHEME, 'Found an invalid notificationEventScheme value')
+
+const getDeployableNotificationEvent = (eventEntries: NotificationEventValue): NotificationEvent[] =>
```
Comment (omrilit) [inline packages/jira-adapter/src/filters/notification_scheme/notification_scheme_deployment.ts:54]: Maybe rename to `toDeployableNotificationEvent`?
I think that when using transformation functions, it's better to use "to" and not "get" in the prefix for better readibility.
If this is the convention in the adapter, you can ignore me here :)

File: packages/jira-adapter/src/filters/notification_scheme/notification_scheme_deployment.ts [hunk 6]
```diff
@@ -13,144 +13,210 @@
...
     }
+  )
+}
+
+const filter: FilterCreator = ({ client, config }) => {
+  const originalChanges: Record<string, Change<InstanceElement>> = {}
+  return {
+    name: 'notificationSchemeDeploymentFilter',
+    onFetch: async (elements: Element[]) => {
+      if (!config.client.usePrivateAPI) {
```
Comment (omrilit) [inline packages/jira-adapter/src/filters/notification_scheme/notification_scheme_deployment.ts:91]: Is it still needed?

File: packages/jira-adapter/src/filters/notification_scheme/notification_scheme_deployment.ts [hunk 7]
```diff
@@ -13,144 +13,210 @@
...
+      )
 
-    const jspRequests = config.apiDefinitions.types[NOTIFICATION_SCHEME_TYPE_NAME]?.jspRequests
-    if (jspRequests === undefined) {
-      throw new Error(`${NOTIFICATION_SCHEME_TYPE_NAME} jsp urls are missing from the configuration`)
-    }
+      const deployResult = await deployChanges(
+        relevantChanges.filter(isInstanceChange),
+        async change => {
+          await defaultDeployChange({
```
Comment (omrilit) [inline packages/jira-adapter/src/filters/notification_scheme/notification_scheme_deployment.ts:165]: Could you please elaborate why we call defaultDeployChange with all of its inner details also when it's a modification? shouldn't we clean some of its inner data before deploying it?

File: packages/jira-adapter/src/filters/account_id/account_id_filter.ts [hunk 8]
```diff
@@ -270,22 +270,28 @@ const filter: FilterCreator = ({ config }) => {
       changes
         .filter(isInstanceChange)
         .filter(isAdditionOrModificationChange)
-        .map(getChangeData)
-        .filter(isDeployableAccountIdType)
-        .forEach(element =>
-          walkOnElement({ element, func: walkOnUsers(cacheAndSimplifyAccountId(cache), config) }))
+        .filter(change => isDeployableAccountIdType(getChangeData(change)))
```
Comment (omrilit) [inline packages/jira-adapter/src/filters/account_id/account_id_filter.ts:273]: Is it relevant to this PR?

File: packages/jira-adapter/test/filters/account_id/account_id_filter.test.ts [hunk 9]
```diff
@@ -174,7 +174,7 @@ describe('account_id_filter', () => {
     it('returns account id structure to a simple and correct string on pre deploy on all 5 types', async () => {
       await filter.preDeploy(displayChanges)
       common.checkSimpleInstanceIds(getChangeData(displayChanges[0]), '0')
-      common.checkObjectedInstanceIds((displayChanges[1] as ModificationChange<InstanceElement>).data.before, '1')
+      common.checkSimpleInstanceIds((displayChanges[1] as ModificationChange<InstanceElement>).data.before, '1')
```
Comment (omrilit) [inline packages/jira-adapter/test/filters/account_id/account_id_filter.test.ts:177]: is it relevant to this PR?

**Review by omrilit — COMMENTED** [review]
Great work!
Had mainly questions inside

File: packages/jira-adapter/src/filters/account_id/account_id_filter.ts (hunk 8 above)
Comment (shir-reifenberg) [reply packages/jira-adapter/src/filters/account_id/account_id_filter.ts:273]: yep, `NotificationSchemeEvents` has type `User` that is being converted with this filter. The issue was we were only transforming back user objects to strings on the `after` part of the change and not the `before` part, and then on deployment of `NotificationSchemeEvents` removals we could not extract the ids for `NotificationSchemeEvents` of type `User` (we need the ID in order to make the `delete` request, and the ID is base on the user string and not the object)
Hopes t's clear enough I can explain IRL

File: packages/jira-adapter/test/filters/account_id/account_id_filter.test.ts (hunk 9 above)
Comment (shir-reifenberg) [reply packages/jira-adapter/test/filters/account_id/account_id_filter.test.ts:177]: same

File: packages/jira-adapter/src/filters/notification_scheme/notification_scheme_deployment.ts (hunk 7 above)
Comment (shir-reifenberg) [reply packages/jira-adapter/src/filters/notification_scheme/notification_scheme_deployment.ts:165]: we clean it using `fieldsToIgnore` in the config (see https://github.com/salto-io/salto/pull/5183/files#diff-30fe015aa09d08972c73cad827947ad3e4d1c41302efc129301dcdc3d57176a7R956)

File: packages/jira-adapter/src/filters/notification_scheme/notification_scheme_deployment.ts (hunk 6 above)
Comment (shir-reifenberg) [reply packages/jira-adapter/src/filters/notification_scheme/notification_scheme_deployment.ts:91]: no, thanks!

File: packages/jira-adapter/src/filters/notification_scheme/notification_scheme_deployment.ts (hunk 4 above)
Comment (shir-reifenberg) [reply packages/jira-adapter/src/filters/notification_scheme/notification_scheme_deployment.ts:62]: it works, but I can remove it

File: packages/jira-adapter/src/filters/notification_scheme/notification_events.ts (hunk 2 above)
Comment (shir-reifenberg) [reply packages/jira-adapter/src/filters/notification_scheme/notification_events.ts:116]: it's a valid usecase, for example on addition change we don't have ids until we deploy the change

File: packages/jira-adapter/src/config/api_config.ts [hunk 10]
```diff
@@ -953,6 +969,25 @@ const DEFAULT_TYPE_CUSTOMIZATIONS: JiraApiConfig['types'] = {
         { fieldName: 'notifications', fieldType: 'List<PermissionHolder>' },
       ],
     },
+    deployRequests: {
```
[the remark this answers is not in the document]
Comment (shir-reifenberg) [reply packages/jira-adapter/src/config/api_config.ts:972]: only additions and removals are supported :(

File: packages/jira-adapter/src/config/api_config.ts [hunk 11]
```diff
@@ -953,6 +969,25 @@ const DEFAULT_TYPE_CUSTOMIZATIONS: JiraApiConfig['types'] = {
...
+      add: {
+        url: '/rest/api/3/notificationscheme/{notificationSchemeId}/notification',
+        method: 'put',
```
[the remark this answers is not in the document]
Comment (shir-reifenberg) [reply packages/jira-adapter/src/config/api_config.ts:975]: [yes](https://developer.atlassian.com/cloud/jira/platform/rest/v3/api-group-issue-notification-schemes/#api-rest-api-3-notificationscheme-id-notification-put) :)

File: packages/jira-adapter/src/filters/notification_scheme/notification_scheme_deployment.ts (hunk 7 above)
Comment (omrilit) [reply packages/jira-adapter/src/filters/notification_scheme/notification_scheme_deployment.ts:165]: Got it, thanks!

**Review by omrilit — APPROVED** [review]
I think that the comments that were left are nits/renamse suggestion etc, approving in advance

**Pull request merged**